### PR TITLE
C3-254: Support for get_network_list call with filter (cluster and host level filter)

### PR DESCRIPTION
**Pull request by avnish30jn**
**Jira ID**

C3-254
C3-253

**Description**
Changes done
Support for get_network_list call with filter (cluster and host level filter) is added
Support for get_datastore_list call which returns the hosts in a cluster is added
Updated the response of get_custer_list, get_datacenter_list and get_host_list call

**Testing**

Done. Logs below:

[root@localhost test]# go run vsphereclient.go
Datastore in a hostsystem :
[{"capacity":"992137445376","freespace":"216217419776","name":"datastore1"},{"capacity":"2000112582656","freespace":"1971742310400","name":"Datastore-ESXi5-2TB"}]

[root@localhost test]# go run vsphereclient.go
Networks in a hostsystem :
[{"name":"VM Network"},{"name":"VMprivate-network"},{"name":"visor_network"}]

[root@localhost test]# go run vsphereclient.go
Networks in a datacenter as per filter :
[{"name":"VM Network"},{"name":"VMprivate-network"},{"name":"visor_network"},{"name":"VM-private-net2"}]

[root@localhost test]# go run vsphereclient.go 
Clusters in a datacenter : 
[{"name":"C3 Cluster-1"}]

[root@localhost test]# go run vsphereclient.go 
Datacenter List :
[{"name":"Jignesh-DataCenter"},{"name":"DETA-Lab"},{"name":"C3 DataCenter"},{"name":"pinkesh-lab"}]

[root@localhost test]# go run vsphereclient.go 
Host in a cluster:
[{"name":"67.220.182.194"},{"name":"67.220.182.195"},{"name":"209.205.216.2"},{"name":"209.205.217.242"}]

### CODE REVIEW

File: virtualmachine/vsphere/vm.go [hunk 1]
```diff
@@ -448,7 +448,6 @@ func (vm *VM) Provision() (err error) {
 			case SKIPTEMPLATE_ERROR:
 				return fmt.Errorf("Template already exists: %s", vm.Template)
 			case SKIPTEMPLATE_OVERWRITE:
-				fmt.Println("Overwriting Template")
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go:451]: why this change is req

File: virtualmachine/vsphere/vm.go [hunk 2]
```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
...
+				return nil, err
+			}
+			networkMap = map[string]string{"name": network.Name}
+		case "DistributedVirtualPortgroup":
+			err = vm.collector.RetrieveOne(vm.ctx, networkMor, []string{"name"}, &portGroup)
+			if err != nil {
+				return nil, err
+			}
+			networkMap = map[string]string{"name": portGroup.Name}
+		case "OpaqueNetwork":
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go:943]: Can we have a default case handling?
At least print the network type, rather than silently ignoring.

File: virtualmachine/vsphere/vm.go [hunk 3]
```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
...
 	}
 	return networkList, nil
 }
 
+// GetDcNetworkList : returns a list of network in given datacenter
+// available-filters (map-keys): "hosts", "clusters".
+func GetDcNetworkList(vm *VM, filter map[string][]string) ([]map[string]string, error) {
+	networkMap := map[string]bool{}
+	networks := []map[string]string{}
+	// set up session to vcenter server
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go]: add blank line above this

File: virtualmachine/vsphere/vm.go [hunk 4]
```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
...
+	}
+
+	vm.Destination.DestinationType = "cluster"
+	for _, cluster := range filter["clusters"] {
+		vm.Destination.DestinationName = cluster
+		for _, host := range filter["hosts"] {
+			vm.Destination.HostSystem = host
+			networksInHost, err := GetNetworkInHost(vm)
+			switch err.(type) {
+			case ErrorObjectNotFound:
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go]: please print an info msg here

File: virtualmachine/vsphere/vm.go [hunk 5]
```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
...
+// GetDcNetworkList : returns a list of network in given datacenter
+// available-filters (map-keys): "hosts", "clusters".
+func GetDcNetworkList(vm *VM, filter map[string][]string) ([]map[string]string, error) {
+	networkMap := map[string]bool{}
+	networks := []map[string]string{}
+	// set up session to vcenter server
+	if err := SetupSession(vm); err != nil {
+		return nil, err
+	}
+	if _, ok := filter["clusters"]; !ok {
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go]: ->
```
if clusters, ok := filter["clusters"]; !ok {
...
```
and use 'clusters` in below code. same for others

File: virtualmachine/vsphere/vm.go [hunk 6]
```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
...
+			}
+			if err != nil {
+				return nil, err
+			}
+			for _, network := range networksInHost {
+				networkMap[network["name"]] = true
+			}
+		}
+	}
+	for k := range networkMap {
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go]: Check if. getting below info in network object make sense
->
```
for _, network := range networkMap {
    networks = append(networks, network)
}
```

Else
```
for k, _ := range networkMap {
...
```

File: virtualmachine/vsphere/vm.go [hunk 7]
```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
...
+	}
+
+	if _, ok := filter["hosts"]; !ok {
+		return nil, errors.New("Key 'hosts' is missing")
+	}
+
+	vm.Destination.DestinationType = "cluster"
+	for _, cluster := range filter["clusters"] {
+		vm.Destination.DestinationName = cluster
+		for _, host := range filter["hosts"] {
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go]: You can first convert all filters to a host-list filter
And then, iterate over that host-list

File: virtualmachine/vsphere/vm.go [hunk 8]
```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
...
+			networksInHost, err := GetNetworkInHost(vm)
+			switch err.(type) {
+			case ErrorObjectNotFound:
+				continue
+			}
+			if err != nil {
+				return nil, err
+			}
+			for _, network := range networksInHost {
+				networkMap[network["name"]] = true
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go]: ->
```
if _, ok := networkMap[network["name"]; !ok {
    networkMap[network["name"]] = network
}
```

**Review by pankajtandon12 — CHANGES_REQUESTED** [review]

File: virtualmachine/vsphere/vm.go (hunk 1 above)
Comment (avnish30jn) [reply virtualmachine/vsphere/vm.go:451]: This Print statement was for testing purpose, removing now.

File: virtualmachine/vsphere/vm.go (hunk 4 above)
Comment (avnish30jn) [reply virtualmachine/vsphere/vm.go]: This is to check if a destination [host in a cluster] exist or not. 'GetNetworkInHost' searches the destination in vcenter server and returned 'ErrorObjectNotFound', if destination does not exist.

**Review by pankajtandon12 — APPROVED** [review]
